tbot/error.py

```diff
@@ -58,3 +85,139 @@ def __init__(self, arg: Any, host: "machine.Machine") -> None:
         self.host = host
 
         super().__init__(f"{arg!r} references a host/machine that is not {host!r}")
+
+
+class ContextError(ApiViolationError, RuntimeError):
+    """
+    A forbidden or wrong kind of interaction with tbot's context was attempted.
+
+    The exception message will include more details.  As an example, this
+    exception is raised when requesting an instance that is currently
+    exclusively requested elsewhere.
+
+    .. versionadded:: UNRELEASED
+    """
+
+
+class MachineNotFoundError(ApiViolationError, IndexError):
+    """
+    The requested machine was not found in the tbot context.
+
+    The most likely cause for this is that the selected configuration did not
+    register a machine for the requested role.
+
+    .. versionadded:: UNRELEASED
+    """
+
+
+_repr = repr
+
+
+class CommandFailure(MachineError):
+    """
+    A command exited with non-zero exit code.
+
+    This exception is raised by ``exec0()`` methods to indicate command failures.
+
+    .. versionadded:: UNRELEASED
+    """
+
+    host: "machine.Machine"
+    """
+    The host where the failed command was executed.
+
+    This host may or may not be available at the time of catching the exception,
+    depending on how far the exception was bubbled up.
+    """
+
+    cmd: Any
+    """
+    The original form of the command, usually a list of arguments.
+
+    For a readable and displayable version, use the ``repr`` field instead.
+    """
+
+    repr: str
+    """
+    String representation of the command which failed.
+
+    Should only be used for diagnostic purposes.
+    """
+
+    def __init__(
+        self, host: "machine.Machine", cmd: Any, *, repr: Optional[str] = None
+    ) -> None:
+        self.host = host
+        self.cmd = cmd
+        if repr is not None:
+            self.cmd_repr = repr
+        else:
+            self.cmd_repr = _repr(self.cmd)
+
+        super().__init__(f"command failed: [{self.host.name}] {self.cmd_repr}")
+
+
+class ChannelClosedError(MachineError):
+    """
+    Error type for exceptions when a channel was closed unexpectedly.
+
+    This error is raised when interaction with a channel is attemped which was
+    either closed explicitly beforehand or which was closed unexpectedly by the
+    remote end.
```

Comment:
I think we might actually be able to differentiate remote closure and "use-after-close" with some tricks in the `ChannelIO`.  I'll keep this in mind for the future, I don't think we should spend any time on this right now...